File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
      int ans = 0;
        int row = matrix.size();
        int col = matrix[0].size();
        vector<vector<int>>dp(row,vector<int>(col,-1));
       
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
               int temp = solve(row,col,i,j,matrix,dp);
                ans = max(ans,temp);
            }
        }
       
        return ans;
    }
    
    
    int solve(int row, int col,int i, int j,vector<vector<int>>&matrix,vector<vector<int>>&dp){
      if(dp[i][j] != -1)return dp[i][j];
       int down = 1;
        int left = 1;
        int right = 1;
        int up = 1;
        //Down Movement
         if(i+1<row && matrix[i+1][j] > matrix[i][j]){
            down = 1 + solve(row,col,i+1,j,matrix,dp);
     
        }
       
        //Left Movement
        if(j-1>=0 &&  matrix[i][j-1]>matrix[i][j]){
            
         left  = 1 + solve(row,col,i,j-1,matrix,dp);
           
        }
       
        //Right Movement
        if(j+1 < col && matrix[i][j+1]>matrix[i][j]){
            
             right = 1 + solve(row,col,i,j+1,matrix,dp);
            
        }
       
        //Up Movement
        if(i-1 >=0 && matrix[i-1][j]>matrix[i][j]){
  
            up = 1 + solve(row,col,i-1,j,matrix,dp);
            
        }
        
        return dp[i][j] = max(up,max(down,max(left,right)));
        
        
    }
};

File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.java
class Solution {
    public  static int longestIncreasingPath(int[][] matrix) {
        int[][] dp = new int[matrix.length][matrix[0].length];
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                int ans = solve(i,j,matrix ,new boolean[matrix.length][matrix[0].length],dp);
                max = Math.max(max,ans);
            }

        }
        return max;


    }


    static int solve(int r,int c,int[][] mat,boolean[][] visit,int[][] dp){

            if(dp[r][c] !=0) return dp[r][c];
        int up = 1;
        int right = 1;
        int down = 1;
        int left = 1;

        visit[r][c] = true;
        if(r-1 >= 0 && !visit[r-1][c]   ){
            if(mat[r-1][c] < mat[r][c])
                up = 1 + solve(r-1,c,mat,visit,dp);
            else up = 1;
        }
        if(c+1 < mat[0].length && !visit[r][c+1]   ){
            if(mat[r][c+1] < mat[r][c])
                right = 1 + solve(r,c+1,mat,visit,dp);
            else right = 1;
        }
        if(r+1 < mat.length && !visit[r+1][c]){
            if(mat[r+1][c] < mat[r][c])
                down = 1 + solve(r+1,c,mat,visit,dp);
            else down = 1;
        }
        if(c-1 >= 0 && !visit[r][c-1]  ){
            if(mat[r][c-1] < mat[r][c])
                left = 1 + solve(r,c-1,mat,visit,dp);
            else left =  1;
        }

        visit[r][c] = false;


        return dp[r][c] = Math.max(up,Math.max(right,Math.max(down,left)));


    }
    static int solve2(int r,int c,int[][] mat,boolean[][] visit,int[][] dp){

        if(dp[r][c] != 0)return dp[r][c];
        int up = 1;
        int right = 1;
        int down = 1;
        int left = 1;

        visit[r][c] = true;
        if(r-1 >= 0 && !visit[r-1][c]   ){
            if(mat[r-1][c] > mat[r][c])
                up = 1 + solve2(r-1,c,mat,visit,dp);
            else up = 1;
        }
        if(c+1 < mat[0].length && !visit[r][c+1]   ){
            if(mat[r][c+1] > mat[r][c])
                right = 1 + solve2(r,c+1,mat,visit,dp);
            else right = 1;
        }
        if(r+1 < mat.length && !visit[r+1][c]){
            if(mat[r+1][c] > mat[r][c])
                down = 1 + solve2(r+1,c,mat,visit,dp);
            else down = 1;
        }
        if(c-1 >= 0 && !visit[r][c-1]  ){
            if(mat[r][c-1] > mat[r][c])
                left = 1 + solve2(r,c-1,mat,visit,dp);
            else left =  1;
        }

        visit[r][c] = false;


        return dp[r][c] = Math.max(up,Math.max(right,Math.max(down,left)));


    }
}